readme.md

```diff
@@ -312,17 +317,14 @@ For a TS example see the following [discussion](https://github.com/pmndrs/zustan
 import { State, StateCreator } from 'zustand'
 import produce, { Draft } from 'immer'
 
-// Immer V8 or lower
-const immer = <T extends State>(
-  config: StateCreator<T, (fn: (draft: Draft<T>) => void) => void>
-): StateCreator<T> => (set, get, api) =>
-  config((fn) => set(produce(fn) as (state: T) => T), get, api)
-
-// Immer V9
-const immer = <T extends State>(
-  config: StateCreator<T, (fn: (draft: Draft<T>) => void) => void>
-): StateCreator<T> => (set, get, api) =>
-  config((fn) => set(produce<T>(fn)), get, api)
+const immer = <T extends State>config: StateCreator<T>): StateCreator<T> => 
```

Comment:
This is not precise. It's previously `StateCreator<T, (fn: (draft: Draft<T>) => void) => void>`.

```suggestion
const immer = <T extends State>(
  config: StateCreator<T,  (fn: ((draft: Draft<T>) => void) | PartialState<T>, replace?: boolean) => void>
): StateCreator<T> => 
```

But, I can't make it work.
https://codesandbox.io/s/nostalgic-grass-rbdu6?file=/src/zustand/middlewares/immer.ts

BTW, you want to try the immer middleware without `pipe` too.

Reply:
Mhh in my point of view the previous typing ist wrong, isn't it? Because `CustomSetState` of the `StateCreator` type doesn't get a `Draft<T>` as first argument. `Draft<T>` is only relevant for the `produce` function of `immer` which is called inside the custom-`set` function.

So the first argument which the `immer`-middleware gets is the `PartialState<T>`. In case that `PartialState<T>` is a function it would be `(state: T) => ...` but not `(state: Draft<T>) => ...`... And inside the middleware the `produce` function is called. Here it has to be cast to `Draft<T>` because the `producer` function of the immer-producer gets a `Draft<T>` as first argument. The `produce` function would return a new immutable state object.

So in case that the first argument (`PartialState<T>`) of the `CustomSetState` is a object we have to return an immutable object too, to satisfy typescript. I found a function exported by the immer library which is called `castImmutable` and seems to do what we need:

```ts
/**
* This function is actually a no-op, but can be used to cast a mutable type
* to an immutable type and make TypeScript happy
* @param value
*/
export declare function castImmutable<T>(value: T): Immutable<T>;
```

So what about the following implementation: https://codesandbox.io/s/laughing-nash-vz53e?file=/src/zustand/middlewares/immer.ts

```ts
import produce, { Draft, castImmutable } from "immer";
import { State, StateCreator } from "zustand";

export const immer = <T extends State>(
  config: StateCreator<T>
): StateCreator<T> => (set, get, api) =>
  config(
    (partial, replace) => {
      const nextState =
        typeof partial === "function"
          ? produce(partial as (state: Draft<T>) => void)
          : castImmutable(partial);
      return set(nextState, replace);
    },
    get,
    api
  );
```

Let me know if I made thinking mistakes.

The point is, that I intend to write middlewares compatible with previous or subsequent middlewars in the chain. If I don't do that, I can't simply compose other middlewares because I have to struggle with the typings.

I know that I can compose immer without `pipe`, but we have other middlewares which we've put in this middleware-chain and want to use them in our custom `create` & `createPersisted` helpers. And because of dry I did't want to attach them in both helpers.

If you agree with the proposed implementation, I would update the pr accordingly. Let me know what your thoughts are.

Reply:
Thanks for your thoughts. Please bear with me.
I know your final goal is to support pipe, but I'd like to stick with immer only typing for now.

More precisely, I would like to allow following example.

```ts
const useStore = create<MyState>(
  immer((set) => ({
    bears: 0,
    increasePopulation: () =>
      set((state) => {
        ++state.bears;
      }),
    removeAllBears: () => set({ bears: 0 })
  }))
);
```

----

Oh, I didn't know
```ts
const useStore = create<MyState>((set) => ({
  bears: 0,
  increasePopulation: () =>
    set((state) => {
      ++state.bears;
    }),
  removeAllBears: () => set({ bears: 0 })
}));
```
this doesn't complain.
So, v3.3.3 complains, but v3.4.0 doesn't.

Okay, this is not your fault... My concern has nothing to do with immer.
Let me investigate this. Give me some time.

wip codesandbox: https://codesandbox.io/s/nervous-nightingale-lv2yc?file=/src/App.tsx

Reply:
Okay, I gave up. Returning `undefined` is actually fine unless `replace=true`.
Types with `replace=true` is not correct at the moment, but fixing it causes too many errors.
As `replace=true` is rare case, let's leave it for now.

In summary, now I understand `PartialState<S>` type is compatible with `(draft: S) => void` since v3.4.0, I will change my mind.